File: ios/Models/ProcessorData.swift
public class ProcessorData: JSONParams {
  private var object: [AnyHashable: Any?]

  override init(target: [AnyHashable: Any]?) {
    self.object = [:]
    
    super.init(target: target)
  }
  
  private func parseJSONToHttpCallInfo(target: [AnyHashable: Any]?) -> NSDictionary? {
    if (target == nil) {
      return nil
    }
    
    let processor = ProcessorHttpCallInfo(target: target)
    return processor.getDict()
  }
  
  private func parseJSONToAdditionalSessionData(target: [AnyHashable: Any]?) -> NSDictionary? {
    if (target == nil) {
      return nil
    }
    
    let processor = ProcessorAdditionalSessionData(target: target)
    return processor.getDict()
  }
  
  private func parseJSONToResult(target: [AnyHashable: Any]?) -> NSDictionary? {
    if (target == nil) {
      return nil
    }
    
    let processor = ProcessorHttpCallInfo(target: target)
    return processor.getDict()
  }
  
  private func parseJSONToIDScanResultsSoFar(target: [AnyHashable: Any]?) -> NSDictionary? {
    if (target == nil) {
      return nil
    }
    
    let processor = ProcessorIDScanResultsSoFar(target: target)
    return processor.getDict()
  }
  
  func parseJSON() {
    let didError = self.getParam("didError", type: Bool.self)
    let responseBlob = self.getParam("responseBlob", type: String.self)
    let httpCallInfo = self.getParam("httpCallInfo", type: [AnyHashable : Any].self)
    let externalDatabaseRefID = self.getParam("externalDatabaseRefID", type: String.self)
    let additionalSessionData = self.getParam("additionalSessionData", type: [AnyHashable : Any].self)
    let result = self.getParam("result", type: [AnyHashable : Any].self)
    let idScanSessionId = self.getParam("idScanSessionId", type: String.self)
    let idScanResultsSoFar = self.getParam("idScanResultsSoFar", type: [AnyHashable : Any].self)
    
    self.object["didError"] = didError
    self.object["responseBlob"] = responseBlob
    self.object["httpCallInfo"] = self.parseJSONToHttpCallInfo(target: httpCallInfo)
    self.object["externalDatabaseRefID"] = externalDatabaseRefID
    self.object["additionalSessionData"] = self.parseJSONToIDScanResultsSoFar(target: additionalSessionData)
    self.object["result"] = self.parseJSONToResult(target: result)
    self.object["idScanSessionId"] = idScanSessionId
    self.object["idScanResultsSoFar"] = self.parseJSONToIDScanResultsSoFar(target: idScanResultsSoFar)
  }
  
  func getDict() -> NSDictionary {
    return NSDictionary(dictionary: self.object as [AnyHashable: Any])
  }
}

File: android/src/main/java/com/azifacemobile/models/ProcessorData.java
package com.azifacemobile.models;

import androidx.annotation.Nullable;

import com.azifacemobile.models.abstracts.JSONParams;
import com.facebook.react.bridge.Arguments;
import com.facebook.react.bridge.ReadableMap;
import com.facebook.react.bridge.WritableMap;

import org.json.JSONObject;

public class ProcessorData extends JSONParams {
  private final WritableMap object;

  public ProcessorData(JSONObject target) {
    super(target);

    this.object = Arguments.createMap();

    this.parseJSON();
  }

  @Nullable
  private WritableMap parseJSONToHttpCallInfo(@Nullable JSONObject target) {
    if (target == null) return null;

    final ProcessorHttpCallInfo processorHttpCallInfo = new ProcessorHttpCallInfo(target);
    return processorHttpCallInfo.getMap();
  }

  @Nullable
  private WritableMap parseJSONToAdditionalSessionData(@Nullable JSONObject target) {
    if (target == null) return null;

    final ProcessorAdditionalSessionData processorAdditionalSessionData = new ProcessorAdditionalSessionData(target);
    return processorAdditionalSessionData.getMap();
  }

  @Nullable
  private WritableMap parseJSONToResult(@Nullable JSONObject target) {
    if (target == null) return null;

    final ProcessorResult processorResult = new ProcessorResult(target);
    return processorResult.getMap();
  }

  @Nullable
  private WritableMap parseJSONToIDScanResultsSoFar(@Nullable JSONObject target) {
    if (target == null) return null;

    final ProcessorIDScanResultsSoFar processorIDScanResultsSoFar = new ProcessorIDScanResultsSoFar(target);
    return processorIDScanResultsSoFar.getMap();
  }

  private void parseJSON() {
    Boolean didError = this.getParam("didError", Boolean.class);
    String responseBlob = this.getParam("responseBlob", String.class);
    JSONObject httpCallInfo = this.getParam("httpCallInfo", JSONObject.class);
    String externalDatabaseRefID = this.getParam("externalDatabaseRefID", String.class);
    JSONObject additionalSessionData = this.getParam("additionalSessionData", JSONObject.class);
    JSONObject result = this.getParam("result", JSONObject.class);
    String idScanSessionId = this.getParam("idScanSessionId", String.class);
    JSONObject idScanResultsSoFar = this.getParam("idScanResultsSoFar", JSONObject.class);
    ReadableMap serverInfo = this.getParam("serverInfo", ReadableMap.class);

    this.object.putBoolean("didError", Boolean.TRUE.equals(didError));
    this.object.putString("responseBlob", responseBlob);
    this.object.putMap("httpCallInfo", this.parseJSONToHttpCallInfo(httpCallInfo));
    this.object.putString("externalDatabaseRefID", externalDatabaseRefID);
    this.object.putMap("additionalSessionData", this.parseJSONToAdditionalSessionData(additionalSessionData));
    this.object.putMap("result", this.parseJSONToResult(result));
    this.object.putString("idScanSessionId", idScanSessionId);
    this.object.putMap("idScanResultsSoFar", this.parseJSONToIDScanResultsSoFar(idScanResultsSoFar));
    this.object.putMap("serverInfo", serverInfo);
  }

  public WritableMap getMap() {
    return this.object.copy();
  }
}
